### PR TITLE
Add SetupSignalHandlerContext

### PR DESCRIPTION
**Pull request by mfojtik**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

The `server.SetupSignalHandler()` provide a channel that users can use to handle SIGTERM and SIGINT signals. This is usually called from `main.go` and propagated down to controller loops and apiserver to handle graceful shutdown.

In some cases, when programs use `context` instead of channels, it is handy to pass the parent context and have the context cancelled when one of the signals above are caught.

```release-note
NONE
```

### CODE REVIEW

**Comment by liggitt** [comment]
is this related to https://github.com/kubernetes/kubernetes/pull/76659 ?

File: staging/src/k8s.io/apiserver/pkg/server/signal.go [hunk 1]
```diff
@@ -44,6 +45,18 @@ func SetupSignalHandler() <-chan struct{} {
 	return stop
 }
 
+// SetupSignalHandlerContext observe SIGTERM and SIGINIT and cancel the passed context
+// when one of the signals is caught.
+// Can only be called once and cannot be called if SetupSignalHandler is already used.
+func SetupSignalHandlerContext(ctx context.Context) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/signal.go]: don't like the signature.

It should rather look like the Golang context wrapper: https://golang.org/pkg/context/#WithTimeout

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/signal.go]: WithSignalHandler(ctx) (ctx, cancel)

Comment (mfojtik) [reply staging/src/k8s.io/apiserver/pkg/server/signal.go]: yeah, I like the WithSignals better, fixed.

**Comment by mfojtik** [comment]
@liggitt it is not, but I dream about world where we stop using stop channels and move to use context.Context everywhere :-)

File: staging/src/k8s.io/apiserver/pkg/server/signal.go [hunk 2]
```diff
@@ -44,6 +45,23 @@ func SetupSignalHandler() <-chan struct{} {
...
 }
 
+// WithSignals is a context.Context wrapper that add SIGTERM and SIGINT handler into parent context
+// passed to this wrapper. When SIGTERM or SIGINT is observed, the context is cancelled.
+func WithSignals(ctx context.Context) (context.Context, context.CancelFunc) {
+	shutdownCtx, cancel := context.WithCancel(ctx)
+	shutDownChan := SetupSignalHandler()
+	go func() {
+		defer func() {
+			if shutdownCtx.Err() != context.Canceled {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/signal.go]: is this necessary?

File: staging/src/k8s.io/apiserver/pkg/server/signal.go [hunk 3]
```diff
@@ -44,6 +45,23 @@ func SetupSignalHandler() <-chan struct{} {
 	return stop
 }
 
+// WithSignals is a context.Context wrapper that add SIGTERM and SIGINT handler into parent context
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/signal.go]: nit: that adds

**Pull request closed without merging**